subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java

```diff
@@ -203,21 +204,31 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
 
         private FileSystemLocationSnapshot snapshotFile(Path absoluteFilePath, String name, BasicFileAttributes attrs) {
             String internedAbsoluteFilePath = intern(absoluteFilePath.toString());
-            HashCode hash = hasher.hash(absoluteFilePath.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
-            FileMetadata metadata = FileMetadata.from(attrs);
-            return new RegularFileSnapshot(internedAbsoluteFilePath, name, hash, metadata);
+            if (attrs.isRegularFile()) {
+                try {
+                    HashCode hash = hasher.hash(absoluteFilePath.toFile(), attrs.size(), attrs.lastModifiedTime().toMillis());
+                    FileMetadata metadata = FileMetadata.from(attrs);
+                    return new RegularFileSnapshot(internedAbsoluteFilePath, name, hash, metadata);
+                } catch (UncheckedIOException e) {
+                    LOGGER.info("Could not read file path '{}'.", absoluteFilePath, e);
+                }
+            }
+            return new MissingFileSnapshot(internedAbsoluteFilePath, name);
         }
 
+        /** unlistable directories (and maybe some locked files) will stop here */
         @Override
         public FileVisitResult visitFileFailed(Path file, IOException exc) {
             // File loop exceptions are ignored. When we encounter a loop (via symbolic links), we continue
             // so we include all the other files apart from the loop.
             // This way, we include each file only once.
             if (isNotFileSystemLoopException(exc)) {
-                String name = file.getFileName().toString();
-                boolean isDirectory = false;
+                String name = intern(file.getFileName().toString());
+                boolean isDirectory = Files.isDirectory(file);
```

Comment:
this will mostly be true from now on since this is mostly called for unreadable directories